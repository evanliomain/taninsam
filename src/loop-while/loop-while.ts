/**
 * @module any=>any
 */
/**
 * This method loop on an input:
 * - at first, output = input
 * - while predicat(output) is true, output = iteratee(output)
 * - finally, returns output
 * @param predicate The function apply on each loop to decide to :
 * continue (return true), or to stop (return false)
 * @param iteratee The iteratee invoked on each loop.
 * @return the function to apply on the input
 * @example
 * ```
 * loopWhile<number>(x => x < 10, x => x ** 2)(2) // 16
 * ```
 * @example Using the chain
 * ```
 * chain(2)
 *   .chain(loopWhile<number>(x => x < 10, x => x ** 2))
 *   .value() // 16
 * ```
 */
export function loopWhile<T>(
  predicate: (element: T) => boolean,
  iteratee: (element: T) => T,
): (input: T) => T {
  return (input: T) => {
    let output = input;
    while (predicate(output)) {
      output = iteratee(output);
    }

    return output;
  };
}
