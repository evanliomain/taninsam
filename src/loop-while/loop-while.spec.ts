import { describe, expect, test } from 'vitest';
import { loopWhile } from './loop-while';

describe('loopWhile function', () => {
  const input = [1, 2, 3, 4, 5];

  test('is a pure function', () => {
    loopWhile<ReadonlyArray<number>>(
      arr => arr.length < 10,
      arr => [...arr, 0],
    )(input);
    expect(input).toMatchSnapshot();
  });

  test('2 |> loopWhile <10,  x => xÂ² === 16', () => {
    expect(
      loopWhile<number>(
        x => x < 10,
        x => x ** 2,
      )(2),
    ).toBe(16);
  });
});
