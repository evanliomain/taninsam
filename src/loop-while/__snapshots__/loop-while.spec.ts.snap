// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loopWhile function > is a pure function 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;
